### ISSUE TITLE
onMapReady not getting called when offline when a (largish) offline region is present

### ISSUE THREAD

**Issue by tnightingale** [mapbox/mapbox-gl-native#4500]
Following up on #4486: Turns out the issue that I encountered only seems to occur when a offline region larger than a particular size has been downloaded. In my test case this occurs with a region of 87.353 megabytes but **not** with a region of 74.243 megabytes.

Testing steps are mostly the same as I outlined in #4486. The difference being that a OfflineRegion of ~87MB (or probably larger) is downloaded the first time application is loaded with the WiFi enabled.

@zugaldia, @jfirebaugh does that make sense?

I am putting together a test Activity in the TestApp to demonstrate.

**Comment by zugaldia** [comment]
@tnightingale Interesting, thanks again for the feedback. Have you noticed by any chance any differences between the first and the second (and further) downloaded regions? In other words, does this behavior happen to all regions downloaded by the app?

Either way, some code / region definitions would be really helpful.

**Comment by tnightingale** [comment]
I have committed my example code here: https://github.com/tnightingale/mapbox-gl-native/commit/a61b04eea62109ad13996565b371ab9fb1e0b3d3.

There is a new Activity called "OnMapLoadBug Map" (for want of a better name). It's a clone of the OfflineActivity but provides two hard-coded OfflineRegionDefinitions instead of providing a method of creating a new one based on the map's viewport.

If you perform the testing steps I defined in #4486 but when loading the map on WiFi instead of just letting the map filling the ambient cache try download the small or large offline region.

I find that if I download the small offline region then restart process with WiFi disabled onMapReady gets called. If I do the same with the large offline region onMapReady never gets called.

The large region is a segment of downtown Wellington, NZ while the small region is just a suburb of the city. I haven't tried with other regions yet.

**Comment by tnightingale** [comment]
As I test this more I am getting inconsistent results: 

I open the application with WiFi disabled, load the map and <code>onMapLoad()</code> will get called as expected. I then kill the application and repeat multiple times in a row and it continues to work. Then seemingly randomly it just stops.

I cannot pinpoint a specific sequence of events or actions that causes it to occur... Though it does seem that the presence of downloaded offline regions increases the chances of it happening.

Once <code>onMapLoad()</code> stops firing however it won't fire again while the device is offline even after killing the application and reopening it. (It continues to work consistently if WiFi is re-enabled but as soon as it is disabled again it will stop). The only way to resolve it seems to be to clear the application's data.

I'm stumped :-\

**Comment by tobrun** [comment]
@tnightingale thank you for contributing the code, I'm trying to reproduce locally.

**Comment by tobrun** [comment]
I was able to identify the following possible scenario's:
 
#### No internet
Having no internet from start results in a black screen without a map ready callback. (works as designed)

![device-2016-03-29-112329](https://cloud.githubusercontent.com/assets/2151639/14103532/ede8506a-f5a0-11e5-8c7c-47c76c4a073f.png)

#### Internet enabled - home - internet disabled - reopen
I had internet enabled at start, opening the Map results in a map ready callback.

![device-2016-03-29-112629](https://cloud.githubusercontent.com/assets/2151639/14103577/1c300cd8-f5a1-11e5-8de6-4f987f2dfb1f.png)

After click home, disable wifi and return to the app. I'm seeing an not reloaded map without a callback. Becuase the MapView wasn't recreated, the map was already ready.

![device-2016-03-29-113058](https://cloud.githubusercontent.com/assets/2151639/14103726/bb204880-f5a1-11e5-96b4-dbabbe68380e.png)

### Internet enabled - back - internet disabled - reopen
I had internet enabled at start, opening the Map results in a map ready callback.
![device-2016-03-29-112629](https://cloud.githubusercontent.com/assets/2151639/14103577/1c300cd8-f5a1-11e5-8de6-4f987f2dfb1f.png)

After I click back, click home, disable wifi, return to the app and reopen OnMapLoadBug. I'm seeing the map reloading with content and seeing the callback occurring.

![device-2016-03-29-113509](https://cloud.githubusercontent.com/assets/2151639/14103843/5086aca2-f5a2-11e5-975d-3f1865450390.png)

### Initial loaded, kill app and reopen

This test will focus on @tnightingale description of `once in offline downloaded and afterwards keep killing app / reopening it` until it stops providing a callback.

In the first tries, I'm able to produce the normal behaviour as shown above but after 4 tries, the callback is not invoked anymore and the camera will not move to the correct position:

![device-2016-03-29-114024](https://cloud.githubusercontent.com/assets/2151639/14103998/0bb6b152-f5a3-11e5-82e9-cbb5661f5b65.png)

**Comment by tobrun** [comment]
@jfirebaugh 

On Android we were waiting for `DID_FINISH_LOADING_MAP` callback from core, it seems this is not called with being offline while having offline enabled. Replacing this event with `DID_FINISH_RENDERING_MAP_FULLY_RENDERED` resolved it for the offline case shown above. I'm going to do some testing to see if this has unwanted side effects. At this point I'm not seeing any. 

FWIW the related code from core:

```cpp
    if (!fullyLoaded) {
        renderState = RenderState::partial;
    } else if (renderState != RenderState::fully) {
        renderState = RenderState::fully;
        view.notifyMapChange(MapChangeDidFinishRenderingMapFullyRendered);
        if (data->loading) {
            data->loading = false;
            view.notifyMapChange(MapChangeDidFinishLoadingMap);
        }
    }
```

We are now waiting for the former notifyMap to occur instead of the latter.

**Comment by jfirebaugh** [comment]
I'm not able to replicate this behavior myself. I'm using the TestApp, simulator, an offline region, some added logging in mbgl core `map.cpp` and `MapView.java`, and turning off wifi, and everything works as expected.

@tobrun Can you post the code and the steps you took to reproduce this? Is it only reproducible with an actual device?

**Comment by tobrun** [comment]
@jfirebaugh 
I was able to reproduce it locally with:
 - Running with internet just once, turn off WiFi
 - Kill app (open task launcher, swipe app out of screen)
 - Open app, go to related activity
   - if not reproducible go back to step 2

I had to go through 4 iterations before being able to reproduce.

**Comment by jfirebaugh** [comment]
Nope, still can't replicate it.

1. Clean build at https://github.com/tnightingale/mapbox-gl-native/commit/4d6a3db232743b4cf5642073085987808c57a2f5
2. With Wifi on, launch TestApp in simulator
3. Go to "OnMapLoadBug Map"
4. Start "Large region" download, wait for completion
5. Turn Wifi off
6. Open task launcher, swipe app out 
7. Relaunch, go to "OnMapLoadBug Map"
8. App and logs indicate onMapReady is called
9. Repeat starting at step 6

I tried steps 6-7 more than a dozen times and in no instance did onMapReady fail to be called.

**Comment by jfirebaugh** [comment]
I was able to replicate the issue after _enabling airplane mode_ in the simulator. This was the key: it does not replicate when turning wifi off in the simulator's host OS.

The issue is a race condition in the state management of `MapData::loading`. When loaded in airplane mode and the style in the offline database:

1. The offline database satisfies the initial request for the style. `MapContext::loadStyleJSON` is triggered, sets `data.loading = true`, and triggers an async update.
2. The revalidation attempt for the style fails due to lack of network access. The request callback runs [this code](https://github.com/mapbox/mapbox-gl-native/blob/0539f7b4cb49e0690b9654f57fe9e2f724e14903/src/mbgl/map/map_context.cpp#L115-L116) which _resets data.loading to false_.
3. The async update runs, triggers a render, and [this code](https://github.com/mapbox/mapbox-gl-native/blob/0539f7b4cb49e0690b9654f57fe9e2f724e14903/src/mbgl/map/map.cpp#L80-L85). The map is fully loaded, and `MapChangeDidFinishLoadingMap` should be emitted, but since `data.loading` is now false, it isn't.

This is a great illustration of why **state is a liability**. Always endeavor to minimize or eliminate state. I'm going to go back and review why `MapData::loading` exists and whether we can eliminate it.

**Comment by tnightingale** [comment]
A race condition would certainly be consistent with my experiences. My other test application seemed to experience the issue more frequently. Which was probably due was due to the fact it's main Activity loaded two maps on screen and called listOfflineRegions() on init.

Great work, thanks for persevering!

**Comment by bleege** [comment]
Wow!  Where you able to replicate this using Airplane Mode on actual device too or just the emulator?  I wonder if on the emulator it was still considering it's Network connection (via the host computer) to be different from it's WiFi connection and therefore still able to load data from the network?  Android Emulator are horrible.

> This is a great illustration of why state is a liability. Always endeavor to minimize or eliminate state.

Truth.  Immutability FTW!

**Comment by jfirebaugh** [comment]
This will affect iOS, because it also uses `MapChangeDidFinishLoadingMap`. Tagging for 3.2.0.

**Comment by jfirebaugh** [comment]
Fixed for iOS with #4524.
